src/main/java/de/bonndan/nivio/assessment/kpi/CustomKPI.java

```diff
@@ -0,0 +1,138 @@
+                        return ((Labeled) component).getLabel(messageLabel);
+                    }
+                    return null;
+                });
+
+        this.label = label;
+        this.ranges = asRanges(ranges);
+        addSpecs(matches);
+    }
+
+    @Override
+    protected List<StatusValue> getStatusValues(String value, String message) {
+
+        List<StatusValue> values = new ArrayList<>();
+        for (Status status : Status.values()) {
+            Optional<Status> statusByRange = getStatusByRange(value);
+            if (statusByRange.isPresent()) {
+                values.add(new StatusValue(label, statusByRange.get(), message));
+                break;
+            }
+
+            if (!matches.containsKey(status)) {
+                continue;
+            }
+            boolean anyMatch = matches.get(status).stream().anyMatch(stringBooleanFunction -> stringBooleanFunction.apply(value));
+            if (anyMatch) {
+                values.add(new StatusValue(label, status, message));
+                break;
+            }
+        }
+
+        return values;
+    }
+
+    private Map<Status, Range<Double>> asRanges(Map<Status, String> ranges) {
+        Map<Status, Range<Double>> rangeMap = new HashMap<>();
+        if (ranges == null) {
+            return rangeMap;
+        }
+
+        ranges.forEach((status, s) -> {
+            String[] split;
+            if (s.contains(SEPARATOR)) {
+                split = s.split(SEPARATOR);
+            } else {
+                split = new String[2];
+                split[0] = s;
+                split[1] = s;
+            }
+            try {
+                rangeMap.put(status, Range.between(Double.valueOf(split[0]), Double.valueOf(split[1])));
+            } catch (NumberFormatException e) {
+                throw new ProcessingException("Failed to parse KPI range: " + s, e);
+            }
+        });
+        return rangeMap;
+    }
+
+    private Optional<Status> getStatusByRange(String value) {
+        if (value == null) {
+            return Optional.empty();
+        }
+
+        double d;
+        try {
+            d = Double.parseDouble(value);
+        } catch (NumberFormatException ignored) {
+            return Optional.empty();
+        }
+        return ranges.entrySet().stream()
+                .filter(entry -> entry.getValue().contains(d))
+                .findFirst()
+                .map(Map.Entry::getKey);
+    }
+
+    private void addSpecs(@Nullable Map<Status, String> statusPatterns) {
+        if (statusPatterns == null)
+            return;
+
+        statusPatterns.forEach((status, strings) -> {
+            List<Function<String, Boolean>> specs = Arrays.stream(strings.split(SEPARATOR)).map(s -> {
+                try {
+                    Pattern p = Pattern.compile(s, Pattern.CASE_INSENSITIVE);
+                    return (Function<String, Boolean>) s1 -> p.matcher(s1).matches();
+                } catch (Exception e) {
```

Comment:
Can we catch a more specific exception here? Can we check before if the pattern can be compiled?

Is the fallback really the desired behavior? What if I supplied an invalid regex (because of a missing brace etc.), but never intended to apply the ``contains`` behavior?

Reply:
What would you suggest? Simply warn the user that the regex is wrong?